google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/wrappers/RowMutation.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.wrappers;
+
+import com.google.api.core.InternalApi;
+import com.google.bigtable.v2.MutateRowRequest;
+import com.google.bigtable.v2.MutateRowsRequest;
+import com.google.bigtable.v2.MutateRowsRequest.Entry;
+import com.google.bigtable.v2.TableName;
+import com.google.cloud.bigtable.data.v2.internal.RequestContext;
+import com.google.protobuf.ByteString;
+import javax.annotation.Nonnull;
+
+public class RowMutation extends AbstractMutation<RowMutation> {
```

Comment:
I feel like we could avoid inheritance here...

Reply:
You are right, I'll refactor it

Reply:
I take that back, I can't figure out a way to avoid inheritence. The mutation DSL needs to be used in 3 places:
- single mutations
- bulk mutations
- the true & false branches of a conditional mutation

Single mutations need: a table, a key & a list of mutations
bulk mutations will need: a key & a list of mutations
conditional mutations need: only a list of mutations

Each concrete mutation is actually a list of mutations for a single row, so it needs to be chainable and return itself. The only way I can think to get rid of inheritance is to add all 3 fields (table, key, mutations) to the dsl and have the leaf create methods leave unused fields blank.

To be honest I don't love either approach, but have a slight preference to keep the existing one to avoid doing more work :)

Please let me know if you have another approach or if you have strong feelings for the other approach

Reply:
So my impression from this implementation is that the only thing that a concrete class like `RowMutation` itself implements is the conversion of an `AbstractMutation` to a `MutateRowsRequest`, and it gets all the data from `AbstractMutation` except for its own fields.

Here is an alternative:

- Rename `AbstractMutation` to `Mutation`
- Create an interface `MutationSerializer` with the following methods:
  - `MutateRowRequest toProto(Mutation mutation, RequestContext requestContext)`
  - `MutateRowRequest toBulkProto(Mutation mutation, RequestContext requestContext)`
- Have `RowMutation.create(...)` construct a `Mutation`, and pass in a `RowMutationSerializer` implementation to it, with the `toProto` and `toBulkProto` implementations
- Have `Mutation` just return `Mutation` from all mutation methods (since nothing in `Mutation` actually needs to be overridden)

In this way, the `toProto` and `toBulkProto` implementations don't need to be public. 

Would this work?

Reply:
I'd like to suggest another option that could accommodate future library additions. It's a little more work up front, but I feel like it could make the code easier to reason about for maintainers than the status quo.

The idea is to do the following:

1. Move all of the `AbstractMutation` implementations into `Mutation`
2. Delete `AbstractMutation`
3. Add an instance of `Mutation` into `RowMutation`
4. Add methods in `RowMutation` to forward to `Mutation`

If you need to support usage of `RowMutation` or `Mutation` in the same place, maybe rename `Mutation` to something more concrete and add a `Mutation` interface.

Reply:
Updated with a compromise:
`AbstractMutation` has been split into `MutationApi` interface and a concrete implementation `Mutation`. `RowMutation` implements `MutationApi` and delegates to an instance of `Mutation`